File: priv/alphonse/lib/list.js
import Tuple from './tuple';
import Atom from './atom';
import Kernel from './kernel';

let List;

List = function(...args){
  if (!(this instanceof List)){
    return new List(...args);
  }

  let _value = Object.freeze(args);

  this.length = function(){
    return _value.length;
  };

  this.get = function(i){
    return _value[i];
  };

  this.value = function(){
    return _value;
  };

  this.toString = function(){
    return _value.toString();
  };

};

List.__MODULE__ = Atom('List');

List.prototype[Symbol.iterator] = function(){
  return this.value()[Symbol.iterator]();
};

List.delete = function(list, item){
  let new_value = [];
  let value_found = false;

  for(let x of list){
    if(x === item && value_found !== false){
      new_value.push(x);
      value_found = true;
    }else if(x !== item){
      new_value.push(x);
    }
  }

  return List(...new_value);
};

List.delete_at = function(list, index){
  let new_value = [];

  for(let i = 0; i < list.length(); i++){
    if(i !== index){
      new_value.push(list.get(i));
    }
  }

  return List(...new_value);
};

List.duplicate = function(elem, n){
  let new_value = [];

  for (var i = 0; i < n; i++) {
    new_value.push(elem);
  }

  return List(...new_value);
};

List.first = function(list){
  if(list.length() === 0){
    return null;
  }

  return list.get(0);
};

List.flatten = function(list, tail = List()){
  let new_value = [];

  for(let x of list){
    if(Kernel.is_list(x)){
      new_value = new_value.concat(List.flatten(x).value());
    }else{
      new_value.push(x);
    }
  }

  new_value = new_value.concat(tail.value());

  return List(...new_value);
};

List.foldl = function(list, acc, func){
  let new_acc = acc;

  for(let x of list){
    new_acc = func(x, new_acc);
  }

  return new_acc;
};

List.foldr = function(list, acc, func){
  let new_acc = acc;

  for (var i = list.length() - 1; i >= 0; i--) {
    new_acc = func(list.get(i), new_acc);
  }

  return new_acc;
};

List.insert_at = function(list, index, value){
  let new_value = [];

  for(let i = 0; i < list.length(); i++){
    if(i === index){
      new_value.push(value);
      new_value.push(list.get(i));
    }else{
      new_value.push(list.get(i));
    }
  }

  return List(...new_value);
};

List.keydelete = function(list, key, position){
  let new_list = [];

  for(let i = 0; i < list.length(); i++){
    if(!Kernel.match__qmark__(list.get(i).get(position), key)){
      new_list.push(list.get(i));
    }
  }

  return List(...new_list);
};

List.keyfind = function(list, key, position, _default = null){

  for(let i = 0; i < list.length(); i++){
    if(Kernel.match__qmark__(list.get(i).get(position), key)){
      return list.get(i);
    }
  }

  return _default;
};

List.keymember__qmark__ = function(list, key, position){

  for(let i = 0; i < list.length(); i++){
    if(Kernel.match__qmark__(list.get(i).get(position), key)){
      return true;
    }
  }

  return false;
};

List.keyreplace = function(list, key, position, new_tuple){
  let new_list = [];

  for(let i = 0; i < list.length(); i++){
    if(!Kernel.match__qmark__(list.get(i).get(position), key)){
      new_list.push(list.get(i));
    }else{
      new_list.push(new_tuple);
    }
  }

  return List(...new_list);
};


List.keysort = function(list, position){
  let new_list = list;

  new_list.sort(function(a, b){
    if(position === 0){
      if(a.get(position).value < b.get(position).value){
        return -1;
      }

      if(a.get(position).value > b.get(position).value){
        return 1;
      }

      return 0;
    }else{
      if(a.get(position) < b.get(position)){
        return -1;
      }

      if(a.get(position) > b.get(position)){
        return 1;
      }

      return 0;
    }

  });

  return List(...new_list);
};

List.keystore = function(list, key, position, new_tuple){
  let new_list = [];
  let replaced = false;

  for(let i = 0; i < list.length(); i++){
    if(!Kernel.match__qmark__(list.get(i).get(position), key)){
      new_list.push(list.get(i));
    }else{
      new_list.push(new_tuple);
      replaced = true;
    }
  }

  if(!replaced){
    new_list.push(new_tuple);
  }

  return List(...new_list);
};

List.last = function(list){
  if(list.length() === 0){
    return null;
  }

  return list.get(list.length() - 1);
};

List.replace_at = function(list, index, value){
  let new_value = [];

  for(let i = 0; i < list.length(); i++){
    if(i === index){
      new_value.push(value);
    }else{
      new_value.push(list.get(i));
    }
  }

  return List(...new_value);
};

List.update_at = function(list, index, fun){
  let new_value = [];

  for(let i = 0; i < list.length(); i++){
    if(i === index){
      new_value.push(fun(list.get(i)));
    }else{
      new_value.push(list.get(i));
    }
  }

  return new_value;
};

List.wrap = function(list){
  if(Kernel.is_list(list)){
    return list;
  }else if(list == null){
    return List();
  }else{
    return List(list);
  }
};

List.zip = function(list_of_lists){
  if(list_of_lists.length() === 0){
    return List();
  }

  let new_value = [];
  let smallest_length = list_of_lists.get(0);

  for(let x of list_of_lists){
    if(x.length() < smallest_length){
      smallest_length = x.length();
    }
  }

  for(let i = 0; i < smallest_length; i++){
    let current_value = [];
    for(let j = 0; j < list_of_lists.length; j++){
      current_value.push(list_of_lists.get(j).get(i));
    }

    new_value.push(Tuple(...current_value));
  }

  return List(...new_value);
};

List.to_tuple = function(list){
  return Tuple.apply(null, list.value());
};

List.append = function(list, value){
  return List(...list.value.slice().push(value));
};

export default List;

File: lib/elixir_script/prelude/list.ex
defmodule ElixirScript.List do

  def duplicate(data, size) do
    do_duplicate(data, size, [])
  end

  defp do_duplicate(_, 0, list) do
    list
  end

  defp do_duplicate(data, size, list) do
    do_duplicate(data, size - 1, list ++ [data])
  end

  def to_tuple(list) do
    JS.new(Elixir.Core.Tuple, list)
  end

  def wrap(list) when is_list(list), do: list
  def wrap(nil), do: []
  def wrap(term), do: [term]

  def append(list, term) do
    concat(list, [term])
  end

  def prepend(list, term) do
    concat([term], list)
  end

  def concat(list_a, list_b) do
    list_a.concat(list_b)
  end

  def first(list) do
    list[0]
  end

  def last(list) do
    list[length(list) - 1]
  end

  def delete(list, item) do
    do_delete(list, item, 0, [])
  end

  defp do_delete(list, item, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      new_list = case list[current_index] do
        ^item ->
          new_list
        _ ->
          new_list ++ [list[current_index]]
      end

      do_delete(list, item, current_index + 1, new_list)
    end
  end

  def delete_at(list, index) do
    do_delete_at(list, index, 0, [])
  end

  defp do_delete_at(list, index, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      new_list = case current_index == index do
        true ->
          new_list
        _ ->
          new_list ++ [list[current_index]]
      end

      do_delete_at(list, index, current_index + 1, new_list)
    end
  end

  def insert_at(list, index, value) do
    do_insert_at(list, index, value, 0, [])
  end

  defp do_insert_at(list, index, value, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      new_list = case current_index == index do
        true ->
          new_list ++ [value, list[current_index]]
        _ ->
          new_list ++ [list[current_index]]
      end

      do_insert_at(list, index, value, current_index + 1, new_list)
    end
  end

  def replace_at(list, index, value) do
    do_replace_at(list, index, value, 0, [])
  end

  defp do_replace_at(list, index, value, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      new_list = case current_index == index do
        true ->
          new_list ++ [value]
        _ ->
          new_list ++ [list[current_index]]
      end

      do_replace_at(list, index, value, current_index + 1, new_list)
    end
  end


  def update_at(list, index, func) do
    do_update_at(list, index, func, 0, [])
  end

  defp do_update_at(list, index, func, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      new_list = case current_index == index do
        true ->
          new_list ++ [func.(list[current_index])]
        _ ->
          new_list ++ [list[current_index]]
      end

      do_update_at(list, index, func, current_index + 1, new_list)
    end
  end


  def foldl(list, acc, func) do
    do_foldl(list, acc, func, [])
  end

  def foldr(list, acc, func) do
    do_foldl(list.concat([]).reverse(), acc, func, [])
  end

  defp do_foldl([], acc, _, new_list) do
    { acc, new_list }
  end

  defp do_foldl(list, acc, func, new_list) do
    { acc, value } = func.(hd(list), acc)
    do_foldl(tl(list), acc, func, new_list ++ [value])
  end

  def flatten(list) do
    do_flatten(list, [])
  end

  def flatten(list, tail) do
    do_flatten(list, []) ++ tail
  end

  defp do_flatten([], flattened_list) do
    flattened_list
  end

  defp do_flatten(list, flattened_list) do
    flattened_list = case hd(list) do
      l when is_list(l) ->
        flattened_list ++ do_flatten(l, [])
      item ->
        flattened_list ++ [item]
    end

    do_flatten(tl(list), flattened_list)
  end


  def keydelete(list, key, position) do
    do_keydelete(list, key, position, [])
  end

  defp do_keydelete([], _, _, new_list) do
    new_list
  end

  defp do_keydelete(list, key, position, new_list) do
    current_value = hd(list)

    new_list = if elem(current_value, position) == key do
      new_list
    else
      new_list ++ [current_value]
    end

    do_keydelete(tl(list), key, position, new_list)
  end

  def keyfind(list, key, position) do
    do_keyfind(list, key, position, nil)
  end

  def keyfind(list, key, position, default) do
    do_keyfind(list, key, position, default)
  end

  defp do_keyfind([], _, _, default) do
    default
  end

  defp do_keyfind(list, key, position, default) do
    current_value = hd(list)

    if elem(current_value, position) == key do
      current_value
    else
      do_keyfind(tl(list), key, position, default)
    end
  end

  def keymember?(list, key, position) do
    keyfind(list, key, position) != nil
  end

  def keyreplace(list, key, position, new_tuple) do
    do_keyreplace(list, key, position, [], new_tuple)
  end

  defp do_keyreplace([], _, _, new_list, _) do
    new_list
  end

  defp do_keyreplace(list, key, position, new_list, new_tuple) do
    current_value = hd(list)

    new_list = if elem(current_value, position) == key do
      new_list ++ [new_tuple]
    else
      new_list ++ [current_value]
    end

    do_keyreplace(tl(list), key, position, new_list, new_tuple)
  end

  def zip(list_of_lists) do
    Elixir.Core.Functions.zip(list_of_lists)
  end
end
